sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveOuterJoin.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.AttributeSet
+import org.apache.spark.sql.catalyst.plans.{LeftOuter, RightOuter}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Join, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * Removes outer join if it only has distinct on streamed side.
+ */
+object RemoveOuterJoin extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case a @ Aggregate(_, _, p @ Project(_, Join(left, _, LeftOuter, _, _)))
```

Comment:
```
== Optimized Logical Plan ==
Aggregate [a#2L], [a#2L]
+- Project [a#2L]
   +- Join LeftOuter, (a#2L = b#6L)
      :- Project [id#0L AS a#2L]
      :  +- Range (0, 200, step=1, splits=Some(2))
      +- Project [id#4L AS b#6L]
         +- Range (0, 300, step=1, splits=Some(2))
```
btw, why do we need `Project` between `Aggregate` and `Join` here?

Reply:
It must has a `Project` between `Aggregate` and `Join` because their output is different if we can remove `Join`.

Reply:
For example, we can not remove Join for the following case:
```sql
SELECT DISTINCT * FROM t1 LEFT JOIN t2 ON a = b

 Aggregate [a#2L, b#6L], [a#2L, b#6L]              
 +- Join LeftOuter, (a#2L = b#6L)                  
    :- Project [id#0L AS a#2L]                     
    :  +- Range (0, 200, step=1, splits=Some(2))   
    +- Project [id#4L AS b#6L]                     
       +- Range (0, 300, step=1, splits=Some(2))   
```

```sql
SELECT DISTINCT b FROM t1 LEFT JOIN t2 ON a = b

Aggregate [b#6L], [b#6L]                             
+- Project [b#6L]                                    
   +- Join LeftOuter, (a#2L = b#6L)                  
      :- Project [id#0L AS a#2L]                     
      :  +- Range (0, 200, step=1, splits=Some(2))   
      +- Project [id#4L AS b#6L]                     
         +- Range (0, 300, step=1, splits=Some(2))   
```

Reply:
Ah, I got it.